### ISSUE TITLE
Clipboard sometimes will not work on Windows if "start in" field is set to a different value

### ISSUE THREAD

**Issue by jsteinaker** [neovim/neovim#7000]
- `nvim --version`: NVIM v0.2.1-371-g48d38d58
- Operating system/version: Windows 8.1 x64

If neovim's (and hence win32yank's) `bin` folder it's not into `PATH` (Windows variable, not neovim's `$PATH` , the clipboard will behave on strange ways.

- `:CheckHealth` may or may not report finding a clipboard tool. Sometimes, if two checks in a row are made, it will report Success on the first and Error on the second, or viceversa.
- The clipboard may or may not work. When it doesn't, it fails with `clipboard: No provider` or `clipboard: Error when invoking provider` messages

On the most common scenario, `:CheckHealth` reports SUCESS finding a clipboard tool but the clipboard doesn't work. If `bin` it's included into `PATH`, then the clipboard works flawlessly and `:CheckHealth` also reports SUCESS.

It seems that `:CheckHealth` and the actual clipboard are using different ways to call `win32yank`. I don't know if being in `PATH` is required, but at least `:CheckHealth` shouldn't report a working clipboard.

**Comment by justinmk** [comment]
Is it possible that your `$PATH` is exceeding 8192 bytes?

> If bin it's included into PATH, then the clipboard works flawlessly and :CheckHealth also reports SUCESS.

On startup nvim appends `bin` to the process path, see `os_setenv_append_path()`.

> If neovim's (and hence win32yank's) bin folder it's not into PATH (Windows variable, not neovim's $PATH , the clipboard will behave on strange ways.

Nvim's `:echo $PATH` should already have `bin` prepended as mentioned above. 

>  If bin it's included into PATH, then the clipboard works flawlessly and :CheckHealth also reports SUCESS.

How are you including it in PATH? Are you modifying `%PATH%` before launching Nvim? Does it work if you modify `$PATH` after launching Nvim, e.g.:

    let $PATH=$PATH.';path/to/bin'

**Comment by jsteinaker** [comment]
>Is it possible that your $PATH is exceeding 8192 bytes?

It's 1173 if we believe an online UTF-8 string size calculator

>Nvim's :echo $PATH should already have bin appended as mentioned above.

Indeed. `bin` is appended, even when I don't specifically set PATH. Also, `:!win32yank` always works.

>How are you including it in PATH? Are you modifying %PATH% before launching Nvim? Does it work if you modify $PATH after launching Nvim, e.g.:
`let $PATH=$PATH.';path/to/bin'`

I use SET on a CMD window, then I launch nvim-qt. If I modify nvim's internal `$PATH` after launch the misbehaviour persists (and an `:echo $PATH` shows `bin` entry repeated, so the variable is being modified).

**Comment by jsteinaker** [comment]
Well, seems it's not related at all with PATH. I had nvim-qt pinned to the taskbar. If I start nvim-qt from here, the problem arises. If I use a normal shortcut or the command line, it doesn't.

I was using the command line for setting PATH, and this led me to the wrong assumption that the problem was there. The fact is that I wasn't using the taskbar shortcut.

**Comment by justinmk** [comment]
It shouldn't be random, still sounds like a bug.

**Comment by TrySound** [comment]
I have same issue. Clipboard often stop working.

**Comment by jsteinaker** [comment]
Agreed. I was tired last night and my head wasn't working properly.

I'll continue to do tests. I'm running nvim-qt without changing the "Start in" option, no issues for the moment, but I'll report back. Also, it's worth noting that I'm trying the 64-bit version of win32yank, and the misbehaviour happens with both this version and the "normal" 32-bit version included with neovim, so at least seems that the 64 bit version is building and working as expected 👍

**Comment by justinmk** [comment]
Is this still a problem?

**Comment by justinmcgrath** [comment]
I have a similar problem. The clipboard breaks after using `:Files` from the fzf.vim plugin. Sometimes it gives the error that win32yank failed, but sometimes it gives no error, but doesn't copy to the clipboard. Sometimes, using `let $path=$path . ';'` allows the clipboard to work again, but sometimes it doesn't.

I'm not sure whether fzf or Neovim has the bug, but using `:Files` with Vim 8.0 doesn't break the clipboard.

**Comment by justinmcgrath** [comment]
`:Files` only seems to break the clipboard if Neovim was started from a shortcut pinned to the taskbar that has a "Start in" directory different from the directory containing nvim-qt.exe.

**Comment by JAffleck** [comment]
FWIW It seems like the clipboard stops working for me if there's a space in the filename path, but when I rename the file I don't have the same issue where it seems to just stop working.

**Comment by huynle** [comment]
For me, it seems to break if I change directory, then do my `:UpdateRemotePlugins` after doing some plugin updates.

It would work if I open up NeoVim, do `:UpdateRemotePlugins` (after plugin updates with Dein), then proceed to changing my directory.

**Comment by justinmk** [comment]
Can anyone reproduce this using the MSVC build? To try the MSVC build, download "Neovim.zip" from one of these links:

- [MSVC win32](https://ci.appveyor.com/project/neovim/neovim/branch/master/job/kgr6oc8u9epd9noy/artifacts)
- [MSVC win64](https://ci.appveyor.com/project/neovim/neovim/branch/master/job/y8r5ug4h6m6u688d/artifacts)

**Comment by justinmcgrath** [comment]
@justinmk, the links do not work for me. I get this error: 

vim-patch:8.0.1151: "vim -c startinsert!" doesn't append (#8886)
Problem:    "vim -c startinsert!" doesn't append.
Solution:   Correct line number on startup. (Christian Brabandt, closes vim/vim#2117)
https://github.com/vim/vim/commit/09ca932f8e7d63a83b39baa7c03d4c6145e3baab
22579
an hour ago by Jan Edmund Lazo (committed by Justin M. Keyes)
master 012cd35b

**Comment by justinmcgrath** [comment]
I couldn't find a 64-bit one, so I downloaded the 32-bit version here: https://ci.appveyor.com/project/neovim/neovim/build/22579/job/xqt5qpjkaxv2s0vp/artifacts

I get the same problem. Copying to the clipboard breaks seemingly randomly.

**Comment by justinmk** [comment]
@justinmcgrath Thanks. That means, at least, someone can debug this using MSVC. 

Building Neovim with MSVC is easy: https://github.com/neovim/neovim/wiki/Building-Neovim#windows--msvc

**Comment by dumblob** [comment]
Any ideas how to avoid addition of the *neovim-qt* bin dir to `%PATH%`? This issue is still present with the latest nightly (0.3.2-597-g8240b8b59).

**Comment by justinmk** [comment]
@dumblob Not sure why you'd want to avoid that, it's how Vim works, and is how`diff.exe`, `cat.exe`, etc. are found.

**Comment by dumblob** [comment]
>@dumblob Not sure why you'd want to avoid that, it's how Vim works, and is how `diff.exe`, `cat.exe`, etc. are found.

Making it portable would help in my case (corporate environment with restrictions).

**Comment by justinmk** [comment]
@dumblob The must be some miscommunication, because it is portable. Nvim adds (on startup) to its process-local `%PATH$` the directory where `nvim.exe` lives. Are you referring to some other directory? And how does that fall afoul of corporate restrictions?

Are you perhaps talking about an installer (I haven't tried it) provided by nvim-qt , or by chocolately or something else?

**Comment by dumblob** [comment]
@justinmk sorry for the confusion - I understood the solution of "adding nvim-qt bin dir to `%PATH%`" as to be required to be done externally (e.g. manually or by some installer) for the user-global `%PATH%`.

@TamaHobbit how did you add the *nvim-qt.exe* bin dir to `%PATH%`?

If nvim adds the `nvim.exe` bin dir to `%PATH%` on its startup, then what's the difference to the solution @TamaHobbit mentioned (`nvim-qt.exe` is in the same dir)? And how come it works for @TamaHobbit? Sounds like magic :wink:.

**Comment by tasawwarH** [comment]
I downloaded neovim (zip version 64 bit) first time on my machine. I had issues with system clipboard on my windows 10 machine (64 bit). I was not able to copy/paste outside content to neovim-qt window. I tried so many things like different version of win32yank (0.0.2, 0.0.3, 0.0.4) but nothing has worked. I noticed my machine has 2008 VC distribution installed on it. I removed it and again tested copy/paste but no luck. I installed 2015 VC version on my machine and everything worked fine for me. Here is my setup details-
1. Neovim's bin is added to PATH
2. init.vim has - set clipboard=unnamedplus
With 2015 VC and above setup I can use nvim-qt system clipboard copy/paste event after setting different "Start In" path. It would be great if someone can answer this question - is required to have to VC on machine in order to get clipboard working?
Note- on my work machine (windows 7 64 bit) Neovim worked without any extra work, just extracted and opened nvim-qt everything worked fine including system clipboard. I just checked 2015VC is installed on that machine machine as well. Unfortunately I cannot uninstall it for testing purpose.

I ma using "nvim-win64.zip" from https://github.com/neovim/neovim/releases/tag/v0.3.1
